### DISCUSSION TITLE
Helm Chart Dead-Locking on Database Migrations

### DISCUSSION THREAD

**Discussion by sbrandtb** [apache/airflow#33888] [Q&A]
Hey,

I want to ask this here before I create a pull request or issue, because it feels like an obvious problem to me and I do not understand why I cannot find any existing issue or discussion on that topic, so I am assuming I am doing something wrong here.

Our issue is that Airflow with default settings basically deadlocks every time a database upgrade (`airflow db migrate`) is required.

By default...

* All pods wait for DB migrations to be finished, i.e. in `webserver.waitForMigrations.enabled=true` (https://github.com/apache/airflow/blob/0cb875b7ec1cebb101866581166cd7b97047f941/chart/values.yaml#L1297)
* DB migration Jobs [use Helm hooks](https://github.com/apache/airflow/blob/0cb875b7ec1cebb101866581166cd7b97047f941/chart/values.yaml#L1090)
* Those hooks are `post-install,post-upgrade`: https://github.com/apache/airflow/blob/0cb875b7ec1cebb101866581166cd7b97047f941/chart/templates/jobs/migrate-database-job.yaml#L46

So, the migrate database job is not run until installation/upgrade is finished (that's the meaning of [post-hooks](https://helm.sh/docs/topics/charts_hooks/#the-available-hooks), right?), which never happens because pods are waiting for the migration before they turn into "Running" status.

Am I missing anything here or are the defaults downright unusable?

I found [this part](https://airflow.apache.org/docs/helm-chart/stable/index.html#installing-the-chart-with-argo-cd-flux-rancher-or-terraform) in documentation that talks about disabling the hooks for ArgoCD and similar, but that does not apply here, does it? I want to use Helm hooks.

Possible solutions that come into my mind:

1. Make the hook `pre-install,pre-upgrade`. I know from experience that this leads to issues if hooks are using resources (ConfigMap, Secret) that are being created by the Chart itself, leading to another chicken-egg problem, but we are not using Helm-managed CMs for this very reason anyway. I would love to overwrite the hook definition, but due to the order how the merge with user defined annotations is done [here](https://github.com/apache/airflow/blob/0cb875b7ec1cebb101866581166cd7b97047f941/chart/templates/jobs/migrate-database-job.yaml#L50), I cannot even overwrite:
    
        airflow:
          migrateDatabaseJob:
            jobAnnotations:
              'helm.sh/hook': "pre-install,pre-upgrade"
              'helm.sh/foo': "bar"

        kind: Job
        metadata:
          annotations:
            # overwrites
            helm.sh/foo: bar
            # still same
            helm.sh/hook: post-install,post-upgrade
            helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded
            helm.sh/hook-weight: "1"

2. Disable `waitForMigrations`, which would lead to temporarily broken deployments between start of new pods and post-hooks. Plus, Airflow might not even start without migrations, leading to the same deadlock again.
3. Run migrations manually in CI before upgrade, but I would rather not go back to writing manual shell scripts.
4. Run migrations in init containers and not hooks, but then again I would rather not go back to writing manual scripts.

So, again... Am I missing anything? Any suggestions?

**Comment by potiuk** [comment]
> I found [this part](https://airflow.apache.org/docs/helm-chart/stable/index.html#installing-the-chart-with-argo-cd-flux-rancher-or-terraform) in documentation that talks about disabling the hooks for ArgoCD and similar, but that does not apply here, does it? I want to use Helm hooks.

This part is precisely about this part afaik. You have not described where you run your Helm and how you deploy it but what you described is wha Argo CD does for example. They run "post-hook" after the pods are already started which is not what regular "helm install" or "helm upgrade" does and it's not according to specification. The "post-hook" is run after the resources have been created but there is no good reason they should wait for the pods to start-up. If you take regular Helm ugprade, the `post-hook` is run in parallel to the pods starting and this what our helm chart does. Argo and some others do more than that and they introduced their own (non-standard) annotations to emulate this behaviour - for some strange reasons - and disable the helm jobs.

This is what helm defines:

```
post-install | Executes after all resources are loaded into Kubernetes
post-upgrade | Executes on an upgrade request after all resources have been upgraded

```

It's not "after all resources are loaded and new containers are successfully started" - ths condition is missing in the "standard" helm chart (and regular chart does not wait for containers when you use standar Helm Chart and Kubernetes).

Or at least this is how I understand that - but if you have other sources of information, I am also all ears. I also find it peculiar Argo and some others are doing it in non-standard way, so maybe there are other reasons for that.

I guess you need to find out what's the equivalent approach for whatever you are using.

The other options are not applicable:
1. pre-install is bad, because we need to have the new containers and airflow code to run the migration. Pre-install runs before the new code is available, so there is no option for that. 
2. Disable wait_for migration is not a good idea because you might accidentally break your database entries when new code will start to use old database. It's unpredictable to see what effects it might have (every migration might be different) so this is no go.
3. Yes. you can do that - shutdown your instance, run migration outside, startup your instance
4. This is bad idea in general because it slows down the startup whenever you restart the container

Generally hooks is a good idea if you either have a regular helm chart, or whatever your Helm "hosting" solution provider suggests as an alternative if they want to delay running the post-hooks.

But maybe you can indeed propose another way that will have no such drawback? We always welcome PRs, but here it seems that:

* we are doing everythign by the book
* someone does it not-by-the-book (but they provide a specialized workaround that you can use instead to do the same)

**Comment by bixel** [comment]
Hey, I seem to have run into the same issue and came to the same conclusions as you @sbrandtb.

I found https://github.com/helm/helm/issues/11778 that could explain the unexpected behaviour. If I get this right, the helm behaviour varies depending on whether the `--wait` flag is used or not: If you `--wait` for an upgrade, the `post-install` hook actually waits for resources to be in some successful state. If you don't `--wait`, resources only have to be loaded.

I guess that would also explain the weird descrepancies between "standard" helm behaviour and ArgoCD, terraform, etc. that you mentioned @potiuk. They probably `--wait` by default.

Since there is a TTL for job cleanup for a while now, I guess the helm hooks could just be dropped to have a consistent behaviour.

**Comment by guhuajun** [comment]
Greetings,

For me, it's a simple solution. I increased the resources limits.

```yaml
migrateDatabaseJob:
  <snipped>

  resources:
    limits:
      cpu: 1000m #128m
      memory: 1024Mi #128Mi
    requests:
      cpu: 100m
      memory: 128Mi
```
